rfcs/20250314-graph-api-bottom-right-causal-mask/README.md

```diff
@@ -0,0 +1,294 @@
+
+```cpp
+row_index_output = pointwise(row_index_output,
+                            std::make_shared<Tensor_attributes>(static_cast<int32_t>(s_kv)),
+                            Pointwise_attributes()
+                                .set_name("row_idx_add_skv")
+                                .set_mode(PointwiseMode_t::ADD)
+                                .set_compute_data_type(DataType_t::INT32));
+```
+
+To support similar usage in oneDNN Graph, extend the `logical_tensor` API as follows:
+
+```cpp
+/// Constructs a logical tensor object with scalar value.
+///
+/// @param tid Logical tensor ID.
+/// @param dtype Elements data type.
+/// @param scalar Scalar value.
+logical_tensor(size_t tid, data_type dtype, void *scalar) {
+    dnnl_graph_logical_tensor_t val;
+    error::wrap_c_api(dnnl_graph_logical_tensor_init_with_scalar(
+                              &val, tid, convert_to_c(dtype), scalar),
+            "could not create logical_tensor with scalar");
+    data = val;
+}
+```
+
+And update the `dnnl_graph_logical_tensor_t` structure to support saving of scalars.
+
+```cpp
+typedef struct {
+    size_t id;
+    int ndims;
+    dnnl_dims_t dims;
+    dnnl_data_type_t data_type;
+    dnnl_graph_tensor_property_t property;
+    dnnl_graph_layout_type_t layout_type;
+    union {
+        dnnl_dims_t strides;
+        size_t layout_id;
+    } layout;
+
+    // New fields for scalar logical tensor support:
+    bool pass_by_scalar;
+    union {
+        float f32_value;
+        int32_t s32_value;
+    } scalar;
+} dnnl_graph_logical_tensor_t;
+```
+
+#### Pros
+
+- No need to create GPU tensors for s_kv and s_q scalars,
+  and other scalars like attn_scale, -inf.
+- Consistent subgraph design with the top-left implicit causal mask.
+- Align with cuDNN Graph's usage.
+
+#### Cons
+
+- Requires extending the Graph API to support scalar logical tensors
+  (an ABI-breaking change that must wait for the oneDNN v4.0 release).
+
+### Option 3: Pass s_kv and s_q by a New Shape Operation
+
+Another option to avoid allocating dedicated GPU tensors is to introduce a
+new Shape operation to extract s_kv and s_q from previous tensor's shape.
+The `Shape` operation generates a 1D shape tensor along specified axes of an
+input tensor. With this operation, s_kv and s_q can be extracted from the last
+two axes of scaled output tensor's shape.
+
+![causal mask subgraph bottom_right with Shape op](images/causal_mask_subgraph_bottom_right_with_shape.png)
+
+#### Shape Operation Specification
+
+##### Attributes
+
+| Attribute Name | Description | Value Type | Supported Values | Required or Optional |
+| :------------- | :---------- | :--------- | :--------------- | :------------------- |
+| [axes] (@ref dnnl::graph::op::attr::axes) | Specifies the axes for generating shape values. | s64 | A s64 list containing the element in the range of [-N, N-1], negative value means counting from the last axis | Required |
```

Comment:
Does the output order of shape relate to the order of axes？ does it support shape.size() larger than original input ndims? does it support scalar shape?